clients/src/test/java/org/apache/kafka/common/message/ProtocolRoundTripConsistencyTest.java

```diff
@@ -0,0 +1,176 @@
+        messageData.setMyNullableString("nullable string");
+        messageData.setMyNullableBytes("nullable bytes".getBytes());
+        messageData.setMyNullableRecords(MemoryRecords.EMPTY);
+        messageData.setMyNullableIntArray(List.of(1, 2, 3));
+
+        nonFlexibleStruct.set("my_nullable_string", "nullable string")
+            .set("my_nullable_bytes", ByteBuffer.wrap("nullable bytes".getBytes()))
+            .set("my_nullable_records", MemoryRecords.EMPTY)
+            .set("my_nullable_int_array", new Object[] {1, 2, 3});
+
+        checkSchemaAndMessageRoundTripConsistency((short) 0, messageData, nonFlexibleStruct);
+    }
+
+    @Test
+    public void testFlexibleWithNullDefault() {
+        messageData.setMyCompactBytes("compact bytes".getBytes());
+        messageData.setMyCompactRecords(MemoryRecords.EMPTY);
+        messageData.setMyCommonStruct(null);
+
+        flexibleStruct.set("my_common_struct", null);
+
+        checkSchemaAndMessageRoundTripConsistency((short) 1, messageData, flexibleStruct);
+    }
+
+    @Test
+    public void testFlexibleWithNonNullValue() {
+        messageData.setMyCompactBytes("compact bytes".getBytes());
+        messageData.setMyCompactRecords(MemoryRecords.EMPTY);
+        messageData.setMyCompactNullableString("compact nullable string");
+        messageData.setMyCompactNullableBytes("compact nullable bytes".getBytes());
+        messageData.setMyCompactNullableRecords(MemoryRecords.EMPTY);
+        messageData.setMyNullableIntArray(List.of(1, 2, 3));
+
+        flexibleStruct.set("my_compact_nullable_string", "compact nullable string")
+            .set("my_compact_nullable_bytes", ByteBuffer.wrap("compact nullable bytes".getBytes()))
+            .set("my_compact_nullable_records", MemoryRecords.EMPTY)
+            .set("my_nullable_int_array", new Object[] {1, 2, 3});
+
+        checkSchemaAndMessageRoundTripConsistency((short) 1, messageData, flexibleStruct);
+    }
+
+    private void checkSchemaAndMessageRoundTripConsistency(short version, AllTypeMessageData message, Struct struct) {
+        ObjectSerializationCache cache = new ObjectSerializationCache();
+        ByteBuffer buf = ByteBuffer.allocate(message.size(cache, version));
+        ByteBufferAccessor serializedMessageAccessor = new ByteBufferAccessor(buf);
+        // Serialize message
+        message.write(serializedMessageAccessor, cache, version);
+
+        ByteBuffer serializedSchemaBuffer = ByteBuffer.allocate(struct.sizeOf());
+        // Serialize schema
+        struct.writeTo(serializedSchemaBuffer);
+
+        assertEquals(serializedSchemaBuffer.position(), serializedMessageAccessor.buffer().position(),
+            "Buffer positions should match between message and schema serialization");
+        assertEquals(serializedSchemaBuffer, serializedMessageAccessor.buffer(),
+            "Buffer contents should be identical between message and schema serialization");
```

Comment:
How about "Generated and non-generated schema should serialize to the same content."?